app/clients/TrackDownloadsClient.ts

```diff
@@ -0,0 +1,254 @@
+        },
+        (e) => {
+          this.handleDownloadFailed(track.trackId, e)
+        },
+      )
+
+      this.cachedDownloadState.set(track.trackId, {
+        status: "DOWNLOADING",
+        progress: currentState.progress,
+        downloadResumable: currentState.downloadResumable,
+      })
+      this.notify(track.trackId)
+    }
+  }
+
+  pauseDownload(trackId: string): void {
+    const currentState = this.fetchDownloadState(trackId)
+    if (currentState.status === "DOWNLOADING") {
+      currentState.downloadResumable.pauseAsync().then(
+        () => {
+          this.cachedDownloadState.set(trackId, {
+            status: "DOWNLOAD_PAUSED",
+            progress: currentState.progress,
+            downloadResumable: currentState.downloadResumable,
+          })
+          this.notify(trackId)
+        },
+        // TODO: Handle this case.
+        (e) => console.log("pause failed", e),
+      )
+    }
+  }
+
+  clearDownload(track: Track): void {
+    FileSystem.deleteAsync(getTrackFileUri(track)).then(
+      () => {
+        this.cachedDownloadState.set(track.trackId, {
+          status: "NOT_DOWNLOADED",
+        })
+        this.notify(track.trackId)
+      },
+      (error) => {
+        console.error(`Failed to delete temporary file of track ${track.trackId}`, error)
+      },
+    )
+  }
+
+  subscribe(trackId: string, callback: () => void): () => void {
+    const entry = this.listeners.get(trackId)
+
+    if (entry) {
+      entry.push(callback)
+    } else {
+      this.listeners.set(trackId, [callback])
+    }
+
+    const unsubscribe = () => {
+      const entries = this.listeners.get(trackId) ?? []
+
+      this.listeners.set(
+        trackId,
+        entries.filter((e) => e !== callback),
+      )
+    }
+
+    return unsubscribe
+  }
+
+  protected notify(trackId: string) {
+    const entries = this.listeners.get(trackId) ?? []
+
+    entries.forEach((callback) => callback())
+  }
+
+  private async handleDownloaded(
+    track: Track,
+    result: FileSystemDownloadResult | null | undefined,
+  ): Promise<void> {
+    // NOTE: The resolved promise equals null on iOS if the download has
+    // been paused despite the type signature of `downloadAsync` only allowing
+    // for `undefined`.
+    if (result === undefined || result === null) {
+      return
+    }
+    __DEV__ && console.log("download successful", result)
+    const destination = getTrackFileUri(track)
+    await FileSystem.moveAsync({ from: result.uri, to: destination })
+
+    // TODO: At this point we drop the DownloadResumable which _should_ also
+    // lead to the callback/subscription stuff being collected by the GC. Do we
+    // need to clean up manually?
+    this.cachedDownloadState.set(track.trackId, {
+      status: "DOWNLOADED",
+      uri: destination,
+      // NOTE: The md5 field always exists since we set the { md5: true }
+      // option above, therefore it's safe to assert `md5 !== undefined` here.
+      md5: result.md5!,
```

Comment:
`!` only tricks the type system into believing it's never `undefined`, the generated JS ends up being simply `md5: result.md5`, so this will blow up at some later state when using `md5`. Good point, I'll add some kind of `assertNotUndefined` utility function which raises here.